text/0005-coretime-interface.md

```diff
@@ -0,0 +1,176 @@
+
+- The interface MUST allow the Relay-chain to be scheduled on a low-latency basis.
+- Individual cores MUST be schedulable, both in full to a single task (a ParaId or the Instantaneous Coretime Pool) or to many unique tasks in differing ratios.
+- Typical usage of the interface SHOULD NOT overload the VMP message system.
+- Worst case usage of the interface MUST NOT cause the Polkadot system to fail.
+- The interface MUST allow for the allocating chain to be notified of all accounting information relevant for making accurate rewards for contributing to the Instantaneous Coretime Pool.
+- The interface MUST allow for Instantaneous Coretime Market Credits to be communicated.
+- The interface MUST allow for the allocating chain to instruct changes to the number of cores which it is able to allocate.
+- The interface MUST allow for the allocating chain to be notified of changes to the number of cores which are able to be allocated by the allocating chain.
+
+## Stakeholders
+
+Primary stakeholder sets are:
+
+- Developers of the Relay-chain core-management logic.
+- Developers of the Brokerage System Chain and its pallets.
+
+_Socialization:_
+
+This content of this RFC was discussed in the Polkdot Fellows channel.
+
+## Explanation
+
+The interface has two sections: The messages which the Relay-chain is able to receive from the allocating parachain (the *UMP message types*), and messages which the Relay-chain is able to send to the allocating parachain (the *DMP message types*). These messages are expected to be able to be implemented in a well-known pallet and called with the XCM `Transact` instruction.
+
+Future work may include these messages being introduced into the XCM standard.
+
+### UMP Message Types
+
+#### `request_core_count`
+
+Prototype:
+
+```
+fn request_core_count(
+    count: u16,
+)
+```
+
+Requests the Relay-chain to alter the number of schedulable cores to `count`. Under normal operation, the Relay-chain SHOULD send a `notify_core_count(count)` message back.
+
+#### `request_revenue_info_at`
+
+Prototype:
+
+```
+fn request_revenue_at(
+    when: BlockNumber,
+)
+```
+
+Requests that the Relay-chain send a `notify_revenue` message back at or soon after Relay-chain block number `when` whose `until` parameter is equal to `when`.
+
+#### `credit_account`
+
+Prototype:
+
+```
+fn credit_account(
+    who: AccountId,
+    amount: Balance,
+)
+```
+
+Instructs the Relay-chain to add the `amount` of DOT to the Instantaneous Coretime Market Credit account of `who`.
+
+It is expected that Instantaneous Coretime Market Credit on the Relay-chain is NOT transferrable and only redeemable when used to assign cores in the Instantaneous Coretime Pool.
+
+#### `assign_core`
+
+Prototype:
+
+```
+type PartsOf57600 = u16;
+enum CoreAssignment {
+    InstantaneousPool,
+    Task(ParaId),
+}
+fn assign_core(
+    core: CoreIndex,
+    begin: BlockNumber,
+    assignment: Vec<(CoreAssignment, PartsOf57600)>,
+    end_hint: Option<BlockNumber>,
```

Comment:
Noting here - the main purpose of the `None` end-hint is that it allows the broker chain to skip sending this message sometimes. 

Some users of #1 will likely split their regions up and defer their `allocate` calls until the last possible moment. When the next-up region has the same assignment as the previous, there's no need to send a new `assign_core` message